README.md

```diff
@@ -0,0 +1,162 @@
+
+```csharp
+[JsonConverter(typeof(SingleValueConverter))]
+public struct AccountId : IEquatable<AccountId>
+{
+    ...
+}
+
+public sealed class Account : Entity<AccountId>
+{
+    public Account(
+        AccountId id,
+        AccountName name,
+        ...)
+        : base(id)
+    {
+        Name = name;
+        ...
+    }
+
+    public AccountName Name { get; }
+
+    ...
+}
+
+[JsonConverter(typeof(SingleValueConverter))]
+public struct TransactionId : IEquatable<TransactionId>
+{
+    ...
+}
+
+public sealed class Transaction : Entity<TransactionId>
+{
+    public Transaction(
+        TransactionId id,
+        AccountId sender,
+        AccountId recipient,
+        ...)
+        : base(id)
+    {
+        Sender = sender;
+        Recipient = recipient;
+        ...
+    }
+
+    public AccountId Sender { get; }
+
+    public AccountId Recipient { get; }
+
+    ...
+}
+```
+
+These types could each have their own repository interfaces, defined within the "Accounting" domain.
+
+```csharp
+public interface IAccountRepository
+{
+    Task<AccountId> Create(Account account);
+
+    Task<Account> Get(AccountId id);
+
+    ...
+}
+
+public interface ITransactionRepository
+{
+    Task<TransactionId> Create(Transaction transaction);
+
+    IEnumerable<Transaction> GetAllSentBy(AccountId accountId);
+
+    ...
+}
+```
+
+The respective implementations of these repositories, potentially defined in a separate "Persistence" layer, would simply be thin wrappers around the `IEntityFacade`.
+
+```csharp
+internal sealed class AccountRepository : IAccountRepository
+{
+    private readonly IEntityFacade _entityFacade;
+
+    public AccountRepository(IEntityFacade entityFacade)
+    {
+        _entityFacade = entityFacade;
+    }
+
+    public async Task<AccountId> Create(Account account)
+    {
+        return (await _entityFacade.Create<Account, AccountId>(account)).Id;
+    }
+
+    public async Task<Account> Get(AccountId id)
+    {
+        return await _entityFacade.Read<Account, AccountId>(id);
+    }
+
+    ...
+}
+
+internal sealed class TransactionRepository : ITransactionRepository
+{
+    private readonly IEntityFacade _entityFacade;
+
+    public TransactionRepository(IEntityFacade entityFacade)
+    {
+        _entityFacade = entityFacade;
+    }
+
+    public async Task<TransactionId> Create(Transaction transaction)
+    {
+        return (await _entityFacade.Create<Transaction, TransactionId>(transaction)).Id;
+    }
+
+    public IEnumerable<Transaction> GetAllSentBy(AccountId accountId)
+    {
+        return _entityFacade.Query<Transaction, TransactionId>()
+                            .Where(t => t.Sender == accountId);
+    }
+
+    ...
+}
+```
+
+These repositories will store their respective entities in a single shared collection, using the entity type names to namespace the IDs and discriminate between each type. Therefore, these names should be considered immutable once entities have been persisted.
```

Comment:
I think we should tone the language down a bit here. In it's current form it sounds like once a developer has named and persisted an entity then they are not able to modify this in future development. In the spirit of DDD and the continual evolution of the ubiquitous language I think we should change this sentence so that it highlights the fact that the entity name is part of the document schema, and would therefore require a data migration if it were changed, but that is does not prohibit refactoring of the domain.